csrc/scheduler/normalization_inner.cpp

```diff
@@ -218,6 +218,36 @@ int64_t getMaxRegisterCountPerThread(
       blocks_per_sm_max * threads_per_block);
 }
 
+// Returns the maximum persistent batch size.
+// For example: assuming we have 64K registers per SM and 28 warps per SM.
+// Each thread can use up to 72 registers. Then minus the register overhead 16,
+// there are 56 registers or 224 bytes to store the persistent buffer.
+// (1) If each reduction element has 1 fp32 buffer and vectorized by 8,
+//     [buffer_bytes_per_batch] = 4 * 8 = 32. Then the maximum persistent
+//      batch size is 224 / 32 = 7
+// (2) If each reduction element has 1 fp16 buffer and vectorized by 8,
+//     [buffer_bytes_per_batch] = 2 * 8 = 16. Then the maximum persistent
+//      batch size is 224 / 16 = 14, which is then capped to
+//      [max_batches_per_block] whose value is 10.
+int64_t getMaxPersistentBatch(
+    const int64_t buffer_bytes_per_batch,
+    const int64_t target_threads_per_sm,
+    const int64_t register_overhead) {
+  // (1) calculate the maximum register count given the target occupancy.
+  int64_t total_register =
+      getRegPerThreadGivenThreadsPerSM(target_threads_per_sm);
```

Comment:
what's the correct way to get current device properties?

Reply:
The scheduler also needs to know which device a given fusion is compiled for since it needs to know hardware properties. I'm also not sure if we are doing the right thing if there're indeed multiple devices of different hardware properties. All of the fusion caching also need to be done for each individual device.

That said, is it common to have a single PyTorch process/thread to manage multiple GPUs? If each process only manages a single device, this won't be an issue.

Reply:
> is it common to have a single PyTorch process/thread to manage multiple GPUs?

I think it is the case with DP... But don't know if people still use that...

With our python API, we certainly can trigger this if we manually switch between devices, we also have all the information to set the right device as well, we just weren't doing that yet. But this might just be a low priority clean up thing.